pkg/metadata/provider_scaleway.go

```diff
@@ -0,0 +1,219 @@
+		return fmt.Errorf("Scaleway: http.NewRequest failed: %s", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	_, err = client.Do(req)
+	if err != nil {
+		return fmt.Errorf("Scaleway: Could not contact state service: %s", err)
+	}
+
+	return nil
+}
+
+// Probe checks if we are running on Scaleway
+func (p *ProviderScaleway) Probe() bool {
+	// Getting the conf should always work...
+	_, err := scalewayGet(scalewayMetadataURL + "conf")
+	if err != nil {
+		log.Printf(err.Error())
+		return false
+	}
+
+	// we are on Scaleway so we need to send a request to tell that the instance has correclty booted
+	err = p.sendBootSignal()
+	if err != nil {
+		log.Printf("Scaleway: Could not signal that the instance booted")
+	}
+
+	return true
+}
+
+// Extract gets both the Scaleway specific and generic userdata
+func (p *ProviderScaleway) Extract() ([]byte, error) {
+	metadata, err := scalewayGet(scalewayMetadataURL + "conf")
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to get conf: %s", err)
+	}
+
+	hostname, err := p.extractInformation(metadata, "hostname")
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to get hostname: %s", err)
+	}
+
+	err = ioutil.WriteFile(path.Join(ConfigPath, Hostname), hostname, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to write hostname: %s", err)
+	}
+
+	instanceID, err := p.extractInformation(metadata, "id")
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to get instanceID: %s", err)
+	}
+
+	err = ioutil.WriteFile(path.Join(ConfigPath, instanceIDFile), instanceID, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to write instance_id: %s", err)
+	}
+
+	instanceLocation, err := p.extractInformation(metadata, "location_zone_id")
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to get instanceLocation: %s", err)
+	}
+
+	err = ioutil.WriteFile(path.Join(ConfigPath, instanceLocationFile), instanceLocation, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to write instance_location: %s", err)
+	}
+
+	publicIP, err := p.extractInformation(metadata, "public_ip_address")
+	if err != nil {
+		// not an error
+		log.Printf("Scaleway: Failed to get publicIP: %s", err)
+	} else {
+		err = ioutil.WriteFile(path.Join(ConfigPath, publicIPFile), publicIP, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("Scaleway: Failed to write public_ip: %s", err)
+		}
+
+	}
+
+	privateIP, err := p.extractInformation(metadata, "private_ip")
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to get privateIP: %s", err)
+	}
+
+	err = ioutil.WriteFile(path.Join(ConfigPath, privateIPFile), privateIP, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("Scaleway: Failed to write private_ip: %s", err)
+	}
+
+	if err := p.handleSSH(metadata); err != nil {
+		log.Printf("Scaleway: Failed to get ssh data: %s", err)
+	}
+
+	// Generic userdata
+	// should use a port below 1024, TODO fix that, getting forbidden for now
```

Comment:
port below 1024 for what?

Reply:
To get scaleway's userdata the source port must be below 1024 (to ensure that the calling has root privileges I guess)

Reply:
One solution could be to call `curl` from the code. Otherwise I guess it is possible to do it purely in Go but I searched and found nothing..

Reply:
oh, just add `CAP_NET_BIND_SERVICE` to the metadata package capabilities in here https://github.com/linuxkit/linuxkit/blob/master/pkg/metadata/build.yml#L13

Reply:
Yep but the main question remains. Should I prefer a pure go implemention, or is it okay to call `curl` from the go code?

Reply:
Go please.

Reply:
its pretty simple eg https://gist.github.com/higebu/96c09c8f03113997fd9a5338ca99764a but you need to change the capabilities or it will fail